### PR TITLE
Fetch images from cloudinary instead of github

### LINKED ISSUE #311: Use Images from Cloudinary instead of Github in the idle member's Page and Assigned Task Page
Use Images from Cloudinary instead of Github in the idle member's Page and Assigned Task Page

### PR DESCRIPTION
**Pull request by heyrandhir**
### What is the change?
Closes #311

- [x] Fetch images from Cloudinary instead of GitHub

### Is it bug?
No

- Steps to repro
- Expected
- Actual

### \*Dev Tested?

- [x] Yes
- [ ] No

#### Platforms

- [x] Web

#### Browsers

- [x] Chrome
- [ ] Safari
- [ ] Firefox

### Before / After Change Screenshots

> For visual or interaction changes. Can be video / screenshot.

### CODE REVIEW

**Comment by rajat-mehra05** [comment]
Hey @heyrandhir Can you also provide which issue no. does this PR closes ? Thanks.

File: src/helperFunctions/getCloudinaryImageUrl.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
+const cloudinaryImageUrl = `https://res.cloudinary.com/realdevsquad/image/upload`;
```
Comment (rajat-mehra05) [inline src/helperFunctions/getCloudinaryImageUrl.ts:1]: Can we follow the convention of naming variables with uppercase ? For example -`DUMMY_IMAGE_URL` which you have earlier used. It's more readable.

File: src/helperFunctions/getCloudinaryImageUrl.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+const userBaseUrl = `https://api.realdevsquad.com`;
```
Comment (rajat-mehra05) [inline src/helperFunctions/getCloudinaryImageUrl.ts:2]: Can we follow the convention of naming variables with uppercase ? For example -DUMMY_IMAGE_URL which you have earlier used.

File: src/helperFunctions/getCloudinaryImageUrl.ts [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+export const getCloudinaryImgURL = async (userName: string,config:string) => {
+    const memberDataUrl = `${userBaseUrl}/users/${userName}`
```
Comment (rajat-mehra05) [inline src/helperFunctions/getCloudinaryImageUrl.ts:5]: you can either use uppercase convention here or make `Url` uppercase like this- `URL`. Not a much of a naming convention but looks better though.

File: src/helperFunctions/getCloudinaryImageUrl.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+    try {
+        let res = await fetch(memberDataUrl);
+        const data = await res.json()
+        const { user } = data;
+        if (user.picture?.publicId) {
+            const imageUrl = `${cloudinaryImageUrl}/${config}/${user.picture.publicId}`
+            return imageUrl;
```
Comment (rajat-mehra05) [inline src/helperFunctions/getCloudinaryImageUrl.ts:12]: Same comments for the naming convention here too. Please resolve the conversations once you are done with all. Thanks.

File: src/helperFunctions/getCloudinaryImageUrl.ts [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+        }
+    } catch (e) {
+        console.error(e)
+    }
+    return `/dummyProfile.png`;
```
Comment (rajat-mehra05) [inline src/helperFunctions/getCloudinaryImageUrl.ts:17]: Why are we not using variable here ?

Comment (heyrandhir) [reply src/helperFunctions/getCloudinaryImageUrl.ts]: Right we should have used it from constants. now changed to use from constants. Thanks.

File: src/helperFunctions/getCloudinaryImageUrl.ts (hunk 4 above)
Comment (heyrandhir) [reply src/helperFunctions/getCloudinaryImageUrl.ts]: Thanks. Done

File: src/helperFunctions/getCloudinaryImageUrl.ts (hunk 3 above)
Comment (heyrandhir) [reply src/helperFunctions/getCloudinaryImageUrl.ts]: Yes `URL` seems better. Modified it now. Thanks.

File: src/helperFunctions/getCloudinaryImageUrl.ts (hunk 2 above)
Comment (heyrandhir) [reply src/helperFunctions/getCloudinaryImageUrl.ts]: Yeah, that would be better. Done that. Thanks

File: src/helperFunctions/getCloudinaryImageUrl.ts (hunk 1 above)
Comment (heyrandhir) [reply src/helperFunctions/getCloudinaryImageUrl.ts]: Yeah, That sounds good. implemented. Thanks.

File: src/components/tasks/card/index.tsx [hunk 6]
```diff
@@ -26,9 +29,28 @@ const Card: FC<Props> = ({
   const statusRedList = [BLOCKED];
   const statusNotOverDueList = [COMPLETED, VERIFIED, AVAILABLE];
   const cardDetails = content;
-  const [assigneeProfilePic, setAssigneeProfilePic] = useState(
-    `${process.env.NEXT_PUBLIC_GITHUB_IMAGE_URL}/${cardDetails.assignee}/img.png`,
-  );
+  const [assigneeProfilePic, setAssigneeProfilePic] = useState(DUMMY_PROFILE_PATH);
+
+  const memoizeAssignProfilePic = useCallback(async () => {
```
Comment (harshith-venkatesh) [inline src/components/tasks/card/index.tsx:34]: If it is repeating , could we think about having a custom hook?

**Review by harshith-venkatesh — DISMISSED** [review]

File: src/components/tasks/card/index.tsx (hunk 6 above)
Comment (heyrandhir) [reply src/components/tasks/card/index.tsx:34]: yeah, It can be done.

**Comment by Maheima** [comment]
@heyrandhir a suggestion: For things like 
Steps to repro
Expected
Actual
before/after screenshots
time
and many points in the PR's description that you have not answered could be removed. These are just pointers to help you write a good description.

File: next.config.js [hunk 7]
```diff
@@ -3,6 +3,6 @@ module.exports = {
     webpack5: true,
   },
   images: {
-    domains: ["raw.githubusercontent.com"],
+    domains: ["raw.githubusercontent.com","res.cloudinary.com"],
```
Comment (ankushdharkar) [inline next.config.js:6]: Should we not remove github content URL, now that we have cloudinary?

Comment (ankushdharkar) [inline next.config.js:6]: Should we not remove github content URL, now that we have cloudinary?

**Review by ankushdharkar — CHANGES_REQUESTED** [review]

File: src/components/tasks/card/index.tsx [hunk 8]
```diff
@@ -26,9 +29,28 @@ const Card: FC<Props> = ({
...
+        if (isComponentMounted) {
+          if (response !== DUMMY_PROFILE_PATH) {
+            setAssigneeProfilePic(response)
+          }
+        }
+    }
+    return ()=>{
+      isComponentMounted = false
+    }
+  },[getCloudinaryImgURL,setAssigneeProfilePic])
```
Comment (ankushdharkar) [inline src/components/tasks/card/index.tsx:48]: This seems like an anti-pattern. Could we dive deeper into not adding these safeguards manually please?

**Review by ankushdharkar — CHANGES_REQUESTED** [review]

File: next.config.js (hunk 7 above)
Comment (heyrandhir) [reply next.config.js:6]: In Challenges, we are using GitHub content URLs. The challenge is not working correctly. I tried looking into it but the data model coming from the backend isn't correct. Hence apart from that, I changed in the remaining places i.e idle members, available tasks etc.

File: src/components/tasks/card/index.tsx (hunk 8 above)
Comment (heyrandhir) [reply src/components/tasks/card/index.tsx]: yeah. I  have fixed it in my latest PR.

**Comment by heyrandhir** [comment]
Converted this to draft as our free usage of Cloudinary has exceded.

**Review by Maheima — CHANGES_REQUESTED** [review]

File: src/components/availability-panel/drag-drop-context/DroppableComponent.tsx [hunk 9]
```diff
@@ -1,23 +1,52 @@
...
     {droppableId === 'tasks' ? (
       <Droppable droppableId="tasks" isCombineEnabled={!isTaskOnDrag}>
         {(provided) => (
           <div ref={provided.innerRef}>
             {unAssignedTasks.map((task, index) => {
               const { id, title } = task;
               return (
-                <DraggableComponent draggableId={id} index={index} title={title} key={id} />
+                <DraggableComponent draggableId={id} index={index} title={title} key={id} imageURL={unAssignedTasksCloudinaryImageLinks[index]}/>
               );
```
[the remark this answers is not in the document]
Comment (heyrandhir) [reply src/components/availability-panel/drag-drop-context/DroppableComponent.tsx:50]: Had missed removing the console log in one place which is fixed now. Thanks.

**Review by Maheima — DISMISSED** [review]
LGTM!

File: src/components/availability-panel/drag-drop-context/DroppableComponent.tsx [hunk 10]
```diff
@@ -1,23 +1,52 @@
...
   idleMembers,
   isTaskOnDrag,
-}) => (
-  <div>
+}) => {
+  const [idleMembersCloudinaryImageLinks,setIdleMembersCloudinaryImageLinks] = useState<string[]>([])
+  const [unAssignedTasksCloudinaryImageLinks,setUnAssignedTasksCloudinaryImageLinks] = useState<string[]>([])
+  const memoizeFetchUsersCloudinaryLinks = useCallback(()=>{
+    if (unAssignedTasks.length > 0) {
+      const allUsersTasks = unAssignedTasks.map(({ id })=>id)
```
[the remark this answers is not in the document]
Comment (heyrandhir) [reply src/components/availability-panel/drag-drop-context/DroppableComponent.tsx]: this was an unwanted change. Hence Reverted it. however here `spread operator` wouldn't serve the purpose.

File: src/components/availability-panel/drag-drop-context/DroppableComponent.tsx [hunk 11]
```diff
@@ -1,23 +1,52 @@
...
+      const lstUsersCldLinks = getAllUsersCloudinaryImageURLLink(allUsersTasks,CLOUDINARY_IMAGE_CONFIGS)
```
[the remark this answers is not in the document]
Comment (heyrandhir) [reply src/components/availability-panel/drag-drop-context/DroppableComponent.tsx]: this was an unwanted change. Hence Reverted it.

**Review by Maheima — DISMISSED** [review]

File: src/components/idleMembers/card/index.tsx [hunk 12]
```diff
@@ -28,7 +39,7 @@ const Card: FC<Props> = ({ idleMemberUserName }) => {
       aria-hidden="true"
     >
       <Image
-        src={isImageAvailable ? `/${DUMMY_PROFILE}` : assigneeProfilePic(idleMemberUserName)}
+        src={isImageAvailable ? `/${DUMMY_PROFILE}` : assigneeProfilePic}
```
Comment (ankushdharkar) [inline src/components/idleMembers/card/index.tsx:42]: Is this sequence looking right? I thought if we have an image available, we want to be showing the `assigneeProfilePic`?

File: src/components/idleMembers/section/index.tsx [hunk 13]
```diff
@@ -1,17 +1,22 @@
 import { FC } from 'react';
 import Card from '@/components/idleMembers/card';
 import classNames from '@/components/idleMembers/section/section.module.scss';
+interface userNameImage  {
```
Comment (ankushdharkar) [inline src/components/idleMembers/section/index.tsx:4]: I find this interface name a little confusing. Is this for the image of the user's name?

File: src/components/idleMembers/card/index.tsx [hunk 14]
```diff
@@ -1,18 +1,29 @@
...
-  idleMemberUserName: string
+  idleMemberUserName: string,
+  imageURL:string
 }
 
 const Card: FC<Props> = ({ idleMemberUserName }) => {
   const [isImageAvailable, setIsImageAvailable] = useState(false);
 
-  const assigneeProfilePic = (name: string) => `${IMAGE_URL}/${name}/img.png`;
+  const userImageUrl = `/dummyProfile.png`
```
Comment (ankushdharkar) [inline src/components/idleMembers/card/index.tsx:17]: Why does this not use the `DUMMY_PROFILE_PATH` constant?

**Review by ankushdharkar — CHANGES_REQUESTED** [review]

File: src/components/idleMembers/section/index.tsx (hunk 13 above)
Comment (heyrandhir) [reply src/components/idleMembers/section/index.tsx]: No, it isn't for the image of the user's name. The interface name was a bit confusing. It's resolved in my latest Commit.

File: src/components/idleMembers/card/index.tsx (hunk 14 above)
Comment (heyrandhir) [reply src/components/idleMembers/card/index.tsx]: Yeah, should have used the constant. Taken care of in the latest commit. Thanks

File: src/components/idleMembers/card/index.tsx (hunk 12 above)
Comment (heyrandhir) [reply src/components/idleMembers/card/index.tsx]: yeah, the sequence was incorrect. Resolved in the latest PR. Thanks

**Comment by heyrandhir** [comment]
Since this is handled in another PR by Sunny hence closing it

**Pull request closed without merging**